rust/cubesql/cubesql/src/compile/rewrite/rules/split/dates.rs

```diff
@@ -136,19 +157,86 @@ impl SplitRules {
            + Sync
            + Send
            + Clone {
-        let granularity_var = var!(granularity_var);
-        let output_granularity_var = var!(output_granularity_var);
+        let date_part_var = var!(date_part_var);
+        let new_date_part_var = var!(new_date_part_var);
         move |_, egraph, subst| {
-            if let Some(ConstantFolding::Scalar(ScalarValue::Utf8(Some(granularity)))) =
-                &egraph[subst[granularity_var]].data.constant
+            // We are going to split DatePart(epoch, ...), leaving original expression inside
+
+            if let Some(ConstantFolding::Scalar(ScalarValue::Utf8(Some(date_part)))) =
+                &egraph[subst[date_part_var]].data.constant
             {
-                if let Some(out_granularity) = parse_granularity_string(&granularity, true) {
-                    let output_granularity = egraph.add(LogicalPlanLanguage::LiteralExprValue(
-                        LiteralExprValue(ScalarValue::Utf8(Some(out_granularity.to_string()))),
-                    ));
-                    subst.insert(output_granularity_var, output_granularity);
-                    return true;
+                let date_part: DatePartToken = if let Ok(date_part) = date_part.parse() {
+                    date_part
+                } else {
+                    return false;
+                };
+                if !matches!(
+                    date_part,
+                    // Julian sates are very similar to epoch, so we handle it here as well
+                    DatePartToken::Special(SpecialTimeUnitToken::Epoch)
+                        | DatePartToken::Special(SpecialTimeUnitToken::Julian)
+                ) {
+                    return false;
                 }
+
+                let new_date_part = egraph.add(LogicalPlanLanguage::LiteralExprValue(
+                    LiteralExprValue(ScalarValue::Utf8(Some(date_part.as_str().to_string()))),
+                ));
+                subst.insert(new_date_part_var, new_date_part);
```

Comment:
We don't actually transform `date_part` here, effectively using it as is. Considering we could get it from constantfolding, we might as well just reuse `?date_part` placeholder, we don't need `?new_date_part` at all.

Reply:
We are changing it: parsing can handle synonyms like `qtr` for `quarter`, but `as_str` always return “standard” representation (I didn’t actually check standard text, only PostgreSQL docs)

Reply:
Ah, I beg your pardon, I missed `date_part` being shadowed on parse. 👍